File: app/src/main/java/com/aexon/widget/AexonRadioButton.java
/*
* Copyright (c) 2026 Fora
* 
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program. If not, see <https://www.gnu.org/licenses/>.
* 
* Contact: Fora <fora060823@gmail.com>
* Created: 27-01-2026
*/
package com.aexon.widget;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.RippleDrawable;
import android.os.Build;
import android.util.AttributeSet;
import android.widget.RadioButton;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import com.aexon.theme.AexonTheme;
import com.aexon.theme.AexonThemeListener;

@RequiresApi(api = Build.VERSION_CODES.O)
public class AexonRadioButton extends RadioButton {
	
	private final AexonThemeListener themeListener = (view, theme) -> applyTheme(theme);
	
	public AexonRadioButton(@NonNull Context context) {
		super(context);
		init();
	}
	
	public AexonRadioButton(@NonNull Context context, @Nullable AttributeSet attrs) {
		super(context, attrs);
		init();
	}
	
	public AexonRadioButton(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
		init();
	}
	
	private void init() {
		applyTheme(AexonTheme.getInstance());
	}
	
	@Override
	protected void onAttachedToWindow() {
		super.onAttachedToWindow();
		AexonTheme.getInstance().addListener(themeListener);
		applyTheme(AexonTheme.getInstance());
	}
	
	@Override
	protected void onDetachedFromWindow() {
		super.onDetachedFromWindow();
		AexonTheme.getInstance().removeListener(themeListener);
	}
	
	private void applyTheme(@NonNull AexonTheme theme) {
		int colorPrimary = theme.getColorPrimary();
		int colorOnSurface = theme.getColorOnSurface();
		int colorOutline = theme.getColorOutline();
		
		setButtonTintList(new ColorStateList(new int[][] {new int[] { -android.R.attr.state_enabled, android.R.attr.state_checked }, new int[] { -android.R.attr.state_enabled, -android.R.attr.state_checked }, new int[] { android.R.attr.state_checked }, new int[] {}}, new int[] {withAlpha(colorPrimary, 0.38f), withAlpha(colorOutline, 0.38f),colorPrimary, colorOutline}));
		applyRippleBackground(colorPrimary, colorOnSurface);
		setTextColor(new ColorStateList(new int[][] {new int[] { -android.R.attr.state_enabled }, new int[] {}}, new int[] {withAlpha(colorOnSurface, 0.38f), colorOnSurface}));
	}
	
	private void applyRippleBackground(int colorPrimary, int colorOnSurface) {
		ColorStateList rippleColor = new ColorStateList(new int[][] {new int[] { android.R.attr.state_checked }, new int[] {}}, new int[] {withAlpha(colorPrimary, 0.12f), withAlpha(colorOnSurface, 0.08f)});
		
		Drawable defaultBg = getBackground();
		if (defaultBg instanceof RippleDrawable) {
			RippleDrawable defaultRipple = (RippleDrawable) defaultBg.mutate();
			defaultRipple.setColor(rippleColor);
			setBackground(defaultRipple);
		} else {
			setBackground(new RippleDrawable(rippleColor, null, null));
		}
	}
	
	private int withAlpha(int color, float alpha) {
		return (color & 0x00FFFFFF) | (Math.round(alpha * 255) << 24);
	}
}

File: app/src/main/java/com/aexon/material/button/AexonRadioButton.kt
package com.aexon.material.button

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.drawable.Drawable
import android.graphics.drawable.RippleDrawable
import android.util.AttributeSet
import android.widget.RadioButton

import com.aexon.annotation.NonNull
import com.aexon.annotation.Nullable
import com.aexon.annotation.RequiresApi
import com.aexon.theme.AexonTheme
import com.aexon.theme.AexonThemeListener
import com.aexon.core.AexonApi.minSdk

class AexonRadioButton : RadioButton {
	
	private val themeListener = AexonThemeListener { _, theme -> applyTheme(theme) }
	
	constructor(@NonNull context: Context) : super(context) {
		init()
	}
	
	constructor(@NonNull context: Context, @Nullable attrs: AttributeSet?) : super(context, attrs) {
		init()
	}
	
	constructor(@NonNull context: Context, @Nullable attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
		init()
	}
	
	private fun init() {
		applyTheme(AexonTheme.getInstance())
	}
	
	override fun onAttachedToWindow() {
		super.onAttachedToWindow()
		AexonTheme.getInstance().addListener(themeListener)
		applyTheme(AexonTheme.getInstance())
	}
	
	override fun onDetachedFromWindow() {
		super.onDetachedFromWindow()
		AexonTheme.getInstance().removeListener(themeListener)
	}
	
	private fun applyTheme(@NonNull theme: AexonTheme) {
		val colorPrimary = theme.colorPrimary
		val colorOnSurface = theme.colorOnSurface
		val colorOutline = theme.colorOutline
		setButtonTintList(ColorStateList(arrayOf(intArrayOf(-android.R.attr.state_enabled, android.R.attr.state_checked), intArrayOf(-android.R.attr.state_enabled, -android.R.attr.state_checked), intArrayOf(android.R.attr.state_checked), intArrayOf()), intArrayOf(withAlpha(colorPrimary, 0.38f), withAlpha(colorOutline, 0.38f), colorPrimary, colorOutline)))
		if (minSdk(RequiresApi.LOLLIPOP)) {
			applyRippleBackground(colorPrimary, colorOnSurface)
		}
		setTextColor(ColorStateList(arrayOf(intArrayOf(-android.R.attr.state_enabled), intArrayOf()), intArrayOf(withAlpha(colorOnSurface, 0.38f), colorOnSurface)))
	}
	
	@RequiresApi(RequiresApi.LOLLIPOP)
	private fun applyRippleBackground(colorPrimary: Int, colorOnSurface: Int) {
		val rippleColor = ColorStateList(
		arrayOf(
		intArrayOf(android.R.attr.state_checked),
		intArrayOf()
		),
		intArrayOf(
		withAlpha(colorPrimary, 0.12f),
		withAlpha(colorOnSurface, 0.08f)
		)
		)
		
		val defaultBg = background
		if (defaultBg is RippleDrawable) {
			val defaultRipple = defaultBg.mutate() as RippleDrawable
			defaultRipple.setColor(rippleColor)
			setBackground(defaultRipple)
		} else {
			setBackground(RippleDrawable(rippleColor, null, null))
		}
	}
	
	private fun withAlpha(color: Int, alpha: Float): Int {
		return (color and 0x00FFFFFF) or (Math.round(alpha * 255) shl 24)
	}
}